data-access-layer/eventmesh/ApiServerClient.js

```diff
@@ -64,6 +67,9 @@ class ApiServerClient {
   init() {
     return Promise.try(() => {
       if (!this.ready) {
+        _.forEach(config.apiserver.crds, crdTemplate => {
```

Comment:
Can we use Promise.map instead of lodash.forEach for concurrency ?

Reply:
+1 for @sauravmndl 's suggestion here. Since a template-compilation is happening synchronously, would prefer if `Promise.map` or `Promise.mapSeries` constructs are brought in here. Pseudo-code for this would look like:

```
return Promise.map(config.apiserver.crds, crdTemplate=>{
  apiserver.addCustomResourceDefinition(....);
}).then(()=>{
return apiserver.loadSpec()
  .then(().....})
```

Reply:
Not needed. It is actually not an async call. The crds are only added in the client. It does not make any call to apiserver. Check this for reference.
https://github.com/godaddy/kubernetes-client/blob/master/examples/using-crds.js#L23

Reply:
@subhankarc : That's true- no actual call to apiserver is made. But the compilation of the template is still happening synchronously through the `addCustomResourceDefinition` function and the `_.template` call. Generally, the template compilation at runtime is a bit of an expensive operation and it's always better to give leeway to other events in the event loop to run.

Reply:
But in that case what would it wait for? `apiserver.addCustomResourceDefinition(....)` is not returning Promise

Reply:
It need not wait for a result. The `apiserver.addCustomResourceDefinition(....)` call would just be made and since it's a blocking call, the event loop would wait for it to complete. In fact, the `map` function also can return nothing since the `Promise.map` call itself would return an _overall Promise_ object that can then be used.

Reply:
Ok. Will change it then.

Reply:
@subhankarc : Check if this works for you:

```
Promise.map(config.apiserver.crds, crdTemplate=>{
   apiserver.addCustomResourceDefinition(yaml.safeLoad(Buffer.from(crdTemplate, 'base64')));
})
.then(()=>{
   return apiserver.loadSpec().......
})

```

Reply:
Sure.

Reply:
Done.